src/Eshopworld.WorkerProcess/WorkerLease.cs

```diff
@@ -93,10 +101,21 @@ internal async Task LeaseAsync()
                 if (CurrentLease != null)
                     OnLeaseAllocated(CurrentLease.LeasedUntil.GetValueOrDefault());
 
-                _timer.Interval =
+                var proposedInterval =
                     CalculateTimerInterval(
```

Comment:
After further investigation, I think the problem is in the fact that we recalculate the interval out of the LeasedUntil value.
The fact we are using floats for calculations, means we can't avoid small drifts in values. In particular, as in the example presented in the bug itself, the LeasedUntil was calculated (for some reason, most likely due to numeric precission) at 11:23:59am but it should have been 11:24:00am (so 1 second too early).

If we accept that such precision issues will keep happening, we cannot assume that we can safely re-calculate the interval out of the LeasedUntil value.

Here is an example:
_____________________

Let's assume we have current time equal to `14:48:01.7765`, what is equivalent to TotalMilliseconds of  `53281000.7765`.
Also, let's assume that our interval is 12 minutes (`720000` milliseconds).

When we attempt to recalculate new interval at the current time, we will do this calculation:
`720000 - ((53281000.7765 / 720000) % 1) * 720000`

what will result in `718999.22350000334`. That means, we will expect to set the timer to wait that many milliseconds before next wake up (all good so far).

So in this example, our LeasedUntil will be calculated by adding current time milliseconds `53281000.7765` and calculated interval milliseconds `718999.22350000334` what will result in `54000000.000000007`, which is `15:00:00`. 

So when we attempt to ___recalculate___ (unnecessarily) the interval moments later, we will do it using the same formula in this (`CalculateTimerInterval`) function:
`720000 - ((54000000.000000007 / 720000) % 1) * 720000`
what will result in 12 minute interval (`719999.99999998976`). 

However, due to float numerics precision, let's assume LeasedUntil was calculated as `53999999.99999999`. In this case, our interval will be calculated by the function incorrectly as: 
`720000 - ((53999999.99999999 / 720000) % 1) * 720000`
what will result in `1.0244548320770264E-08` (a tiny fraction of time and simulates exactly the reported problem).

If we used the interval calculated originally (i.e. `718999.22350000334`) while trying to derive LeasedUntil, we would not have to recalculate it again (which comes as a different number anyway) and everything would work as expected.

So the solution, in my opinion, is to store the calculated interval in the ILease object, along side of the LeasedUntil value, and use that calculated interval value when setting the timer, instead of recalculating it again.

Reply:
Thanks @chomick, I will look at implementing your suggestion and run some tests